### ISSUE TITLE
Using ground TIN for calculation of Z values

### ISSUE THREAD

**Issue by smgrz** [PDAL/PDAL#2468]
Hi,
it would be nice to have an additional option for filters.hag, to choose whether NN (Nearest Neighbor) or TIN (TIN - Triangulated Irregular Network like in LAStools http://lastools.org/download/lasheight_README.txt) is used for calculation of Z values.
In the current implementation default is NN, there are no further options, am I right? Using NN especially in hilly areas provides a jagged Z.

**Comment by chambbj** [comment]
That's correct, the current implementation is quite limited. I'm sure we would happily accept a PR to expose additional options you may need, if you'd be willing to provide. Another thing we've considered is to compute against externally provided DEMs. Furthermore, I believe @gadomski also has an idea or two in the works that would make this all more robust.

**Comment by gadomski** [comment]
Yeah, I've got some [work in progress](https://github.com/gadomski/PDAL/tree/hag-improved) that improves (at least hopefully) `filters.hag`. Specifically, my improved filter

1. uses multiple nearby points instead of one, interpolating a ground height using inverse distance weighting;
2. checks for "extrapolation", aka when the bounding box of the nearest neighbors does not include the sample point; and
3. includes a "maximum distance" parameter that excludes a non-ground point from HAG calculation if the nearest neighbors are too far away.

The end effect is a `filters.hag` that behaves somewhat like an "adaptive DEM" calculation by (more or less) calculating a grid cell height underneath each non-ground point.

Using a TIN would be another valid approach, but I didn't go with that simply because PDAL doesn't currently have a built-in 3D TIN, so improving the HAG as I did was the lowest-impact solution I could think of. I'd say having both approaches wouldn't be bad, if someone wanted to add the TIN one. I'm going to just implement the improved HAG for now.

**Comment by abellgithub** [comment]
What does a good TIN look like?  We have some surface reconstruction algorithms, but I'm not sure that's what we want and they're really slow.  TINning trees and such seems not worthwhile.

**Comment by smgrz** [comment]
@gadomski it sounds very good to have some additional interpolation and weighting for Nearest Neighbours (NN).

I wrote a python script to calculate (interpolate) Z based on TIN (created from ground points). 
The script is maybe not well readable, because I tried to write it more efficiently. For large point clouds it is not memory efficient! 
Tested with: windows 10, python 3.7, numpy 1.16.2, scipy 1.2.1, sklearn 0.20.3, pdal 1.9.0

To use this script you have to save it with “.py” extension and call it from a pdal pipeline (see examples below).

```python
# -*- coding: utf-8 -*-
import numpy as np
from scipy.spatial import Delaunay
from sklearn.neighbors import BallTree 
import pdal

def z_interpolation(tri, ind_simplex, data_arr, ground_pts_ind, non_ground_pts_ind, ind_pts, ind_p_simpl, ind_p_simpl_outs, InterpMethod):
    '''
    InterpMethod = 'voronoi'
    InterpMethod = 'w_average' 
    InterpMethod = 'barycentric'
    see: https://codeplea.com/triangular-interpolation
    '''
   
    z_interpolated = np.zeros( len(data_arr) )
    
    if InterpMethod=='voronoi':
        # z of ground points (no need for interpolation)
        z_interpolated[ground_pts_ind] = data_arr[ground_pts_ind,2]
        
        # z of non ground points & inside Delaunay
        diff = data_arr[ground_pts_ind][ind_pts[:,:],:2]-data_arr[non_ground_pts_ind[ind_p_simpl],:2][:,None]
        argmin = np.argmin(np.sqrt(np.einsum('ijk,ijk->ij',diff,diff)), axis=1)
        z_interpolated[non_ground_pts_ind[ind_p_simpl]] = data_arr[ground_pts_ind][np.take_along_axis(ind_pts,argmin[:,None], axis=1),2][:,0]
        
        # z of non ground points & outside Delaunay: take z value from ground points' NN
        if len(ind_p_simpl_outs)>0:
            balltree = BallTree( data_arr[ground_pts_ind,:2] )
            nn_index = balltree.query( data_arr[non_ground_pts_ind[ind_p_simpl_outs],:2], k=1, return_distance=False )
            z_interpolated[non_ground_pts_ind[ind_p_simpl_outs]] = data_arr[ground_pts_ind[nn_index[:,0]],2]

    elif InterpMethod=='w_average':
        # z of ground points (no need for interpolation)
        z_interpolated[ground_pts_ind] = data_arr[ground_pts_ind,2]

        # z of non ground points & inside Delaunay
        diff = data_arr[ground_pts_ind][ind_pts[:,:],:2]-data_arr[non_ground_pts_ind[ind_p_simpl],:2][:,None]
        dist = np.sqrt(np.einsum('ijk,ijk->ij',diff,diff))
        bool_notZero = dist == 0
        dist = np.divide(1.0, dist, out=np.zeros_like(dist), where=dist!=0) # inverse dist: 1.0/dist
        z_interpolated[non_ground_pts_ind[ind_p_simpl]] = np.sum(np.einsum('ij,ij->ij', data_arr[ground_pts_ind][ind_pts,2], dist), axis=1) / np.sum(dist,axis=1)
        z_interpolated[non_ground_pts_ind[ind_p_simpl[np.where(bool_notZero)[0]]]] = data_arr[ground_pts_ind][ind_pts,2][bool_notZero] # for all dist==0 correct z value

        # z of non ground points & outside Delaunay: take z value from ground points' NN
        if len(ind_p_simpl_outs)>0:
            balltree = BallTree( data_arr[ground_pts_ind,:2] )
            nn_index = balltree.query( data_arr[non_ground_pts_ind[ind_p_simpl_outs],:2], k=1, return_distance=False )
            z_interpolated[non_ground_pts_ind[ind_p_simpl_outs]] = data_arr[ground_pts_ind[nn_index[:,0]],2]

    elif InterpMethod=='barycentric':
        
        X_p = tri.transform[ind_simplex,:2]
        Y_p = data_arr[non_ground_pts_ind][:,:2]  - tri.transform[ind_simplex,2]
        b = np.einsum('ijk,ik->ij', X_p, Y_p)
        bcoords = np.c_[b, 1 - b.sum(axis=1)]
        np.round(bcoords,2)

        # z of ground points (no need for interpolation)
        z_interpolated[ground_pts_ind] = data_arr[ground_pts_ind,2]

        # z of non ground points & inside Delaunay
        # #short form has ERROR!!!! z_interpolated[non_ground_pts_ind[ind_p_simpl]] = np.sum( np.einsum( 'ij,ij->ij', data_arr[ground_pts_ind][ind_pts,2], bcoords[ind_p_simpl]), axis=1 )
        pt1 = data_arr[ground_pts_ind][ind_pts[:,0],2] * bcoords[ind_p_simpl,0]
        pt2 = data_arr[ground_pts_ind][ind_pts[:,1],2] * bcoords[ind_p_simpl,1]
        pt3 = data_arr[ground_pts_ind][ind_pts[:,2],2] * bcoords[ind_p_simpl,2]
        P = pt1+pt2+pt3
        z_interpolated[non_ground_pts_ind[ind_p_simpl]] = P
    
        # z of non ground points & outside Delaunay: take z value from ground points' NN
        if len(ind_p_simpl_outs)>0:
            balltree = BallTree( data_arr[ground_pts_ind,:2] )
            nn_index = balltree.query( data_arr[non_ground_pts_ind[ind_p_simpl_outs],:2], k=1, return_distance=False )
            z_interpolated[non_ground_pts_ind[ind_p_simpl_outs]] = data_arr[ground_pts_ind[nn_index[:,0]],2]
    
    else:
        print('InterpMethod "{}" unknown!'.format(InterpMethod))
        
    return z_interpolated


def Read_GroundData(in_FilePath):
    # Reads Point Cloud or Raster Data Files 
    if '\\'in in_FilePath:
        print('Please use forward "/" slashes!')
        in_FilePath = in_FilePath.replace('\\','/')
    pl = '{"pipeline":["'+in_FilePath+'"]}'
    pipeline = pdal.Pipeline(pl)
    count = pipeline.execute()
    arrays = pipeline.arrays
    return arrays

def hag_TIN(ins, outs):

    InterpMethod = 'barycentric' # InterpMethod = 'voronoi' or 'w_average' or 'barycentric'  see: https://codeplea.com/triangular-interpolation
    GroundFile = None
    NoDataValue = -9999
    ClassValue = 2 # -1 use all points
    
    # Prepare x,y,z as numpy array
    data = np.array([ins['X'], ins['Y'], ins['Z']]).T
    # Get defined Parameter/Objects
    if 'InterpMethod' in pdalargs.keys():
        InterpMethod = str(pdalargs['InterpMethod'])
    if 'ClassValue' in pdalargs.keys():
        ClassValue = pdalargs['ClassValue']
    if 'NoDataValue' in pdalargs.keys():
        NoDataValue = float(pdalargs['NoDataValue'])
    if 'GroundFile' in pdalargs.keys():
        # Read Ground File
        In_GroundFilePath = pdalargs['GroundFile']
        ground_pts = Read_GroundData(In_GroundFilePath)[0]
        if 'Z' in ground_pts.dtype.names: # LAS/LAZ
            if ClassValue == -1: # -1 --> Use all points
                ground_pts = np.array( [ground_pts['X'],ground_pts['Y'],ground_pts['Z']]).T
            else: # Use points with CalssValues
                bool_ground = ground_pts['Classification'] == ClassValue # Extract only ground points
                ground_pts = np.array( [ground_pts['X'],ground_pts['Y'],ground_pts['Z']]).T
                ground_pts = ground_pts[bool_ground,:]
        else: # Raster
            ground_pts = np.array( [ground_pts['X'],ground_pts['Y'],ground_pts[ground_pts.dtype.names[2]]]).T
            # Eliminate no data raster cells 
            ground_pts = ground_pts[ground_pts[:,2]!=NoDataValue]
        # Get indices of ground and non ground points
        ground_pts_ind = np.arange(len(ground_pts))
        non_ground_pts_ind =np.arange(len(ground_pts), len(ground_pts)+len(data) )
        # Prepare numpy array
        data_arr = np.append(ground_pts,data, axis=0)
    else:
        # Get indices of ground and non ground points
        ground_pts_ind =  np.where(ins['Classification']==ClassValue)[0]
        non_ground_pts_ind = np.where(ins['Classification']!=ClassValue)[0]
        # Prepare numpy array
        data_arr = data

    # Reduce x,y values (only temporary , due to precision!)
    data_arr[:,0] = data_arr[:,0] - np.min( data_arr[:,0] )
    data_arr[:,1] = data_arr[:,1] - np.min( data_arr[:,1] )
    
    # Calculate Delaunay
    tri = Delaunay(data_arr[ground_pts_ind,:2])
    simplices = tri.simplices 
    
    # Find simplex for non ground points
    ind_simplex = tri.find_simplex( data_arr[non_ground_pts_ind,:2] )
    
    # Points connected to siplices
    bool_simpl = ind_simplex>-1
    ind_p_simpl = np.where(bool_simpl)[0] # indices of points witch belogs to a triangle
    ind_p_simpl_outs = np.where(np.invert(bool_simpl))[0]  # indices of points outside of Delaunay triangulation
    
    # Point indices of Delaunay triangles
    ind_pts = simplices[ind_simplex[ind_p_simpl]]
    
    # Interpolate z values
    z_interpolated = z_interpolation(tri, ind_simplex, data_arr, ground_pts_ind, non_ground_pts_ind, ind_pts, ind_p_simpl, ind_p_simpl_outs, InterpMethod=InterpMethod )
    
    # Calculate HeightAboveGround (normalize)
    z_normed = data_arr[:,2] - z_interpolated
    
    if 'GroundFile' in pdalargs.keys():
        z_normed = z_normed[non_ground_pts_ind]
    
    outs['HeightAboveGround'] = z_normed
    return True
```

Following are examples to call "hag_TIN" function from python and calculate HeightAboveGround:
Interpolation Methods = "barycentric" or "w_average" or "voronoi" see: https://codeplea.com/triangular-interpolation

Example 1: Use Ground Points from Input File "in_lasfile.las" classified as 2 (default) or pass through "ClassValue" in "pdalargs": {"InterpMethod": "barycentric","ClassValue":3} other class value
```python
pdal_hag_TIN_pipeline_example = '''[
    "C:/path/to/in_lasfile.las",
    {
        "type": "filters.ferry",
        "dimensions": "=>HeightAboveGround"
    },
    {
        "type":"filters.python",
        "module":"TIN",
        "function":"hag_TIN",
        "script":"C:/path/to/python_FileScript/hag_basedOnTIN.py",
        "pdalargs": {"InterpMethod": "barycentric"}
    },
    {
        "type":"filters.ferry",
        "dimensions":"HeightAboveGround=Z"
    },
    {
        "type":"writers.las",
        "filename":"C:/path/to/out_lasfile.las"
    }
]'''
import pdal
pipeline = pdal.Pipeline(pdal_hag_TIN_pipeline_example)
pipeline.execute()
```

Example 2: Read ground points from another ".las" file and use only points with classification "2"
```python
    {
        "type":"filters.python",
        "module":"TIN",
        "function":"hag_TIN",
        "script":"C:/path/to/python_FileScript/hag_basedOnTIN.py",
        "pdalargs": {"InterpMethod":"barycentric", "GroundFile":"C:/path/to/in_groundFile.las", "ClassValue":2}
    },
```
Example 3: Read ground points from a geo raster (GeoTiff) file and filter out no data values
```python
    {
        "type":"filters.python",
        "module":"TIN",
        "function":"hag_TIN",
        "script":"C:/path/to/python_FileScript/hag_basedOnTIN.py",
        "pdalargs": {"InterpMethod":"barycentric", "GroundFile":"C:/path/to/in_rasterGround.tif", "NoDataValue":-9999}
    },
```

**Comment by abellgithub** [comment]
I guess you've assumed my question in your answer -- you only want a TIN of ground points.  Are there other cases as well?

**Comment by smgrz** [comment]
Yes, I want only TIN of ground points. 
Sometimes I’m using 3D TIN (LAStools: lasview.exe) only for visual identification of outliers, other cases are unknown to me.

**Comment by jamesmcclain** [comment]
I took a pass at adding TIN support in [this branch](https://github.com/jamesmcclain/PDAL/tree/delaunay-fan).

I am new to this community, so I do not know what the expectations are in terms of speed, robustness, and coding style.  For that reason, this branch takes a straight-forward, no-additional-dependencies approach.  The branch is built on top of a rebased version of @gadomski 's [nearest neighbor branch](https://github.com/gadomski/PDAL/tree/hag-improved).

Rather than explicitly constructing a TIN, I have taken the approach of locally calculating the Delaunay neighborhood at each point, which should be equivalent (albeit with a parameter, the neighborhood size).  I assume that the projection (x, y, z) -> (x, y) is one-to-one and use that assumption to project then triangulate neighborhoods of ground points.  Using that assumption a second time, the height of each point is calculated by projecting it along z onto the surface.

I have left this as just a branch rather than a WiP PR because it is built on top of another branch without an associated PR.

I am eager to receive any comments that may be forthcoming.

Input:
```json
[
    "autzen.laz",
    {
        "type":"filters.decimation",
        "step": 20
    },
    {
        "type":"filters.hag",
	"count": 30,
	"delaunay_fans": true
    },
    {
      "type":"writers.gdal",
      "filename":"autzen-hag-tin.tif",
      "dimension":"HeightAboveGround",
      "data_type":"double",
      "output_type":"mean",
      "resolution":10.0
    }
]
```

Output:

![Screenshot from 2019-05-15 21-15-04](https://user-images.githubusercontent.com/11281373/57819942-a58fc180-7758-11e9-942d-6ebd7bfd29ac.png)

**Comment by abellgithub** [comment]
Go ahead and make a PR. Thanks!

**Comment by jamesmcclain** [comment]
> Go ahead and make a PR. Thanks!

https://github.com/PDAL/PDAL/pull/2520

**Comment by AndrewCunliffe** [comment]
FWIW, I'd also vote in favour of supporting computing dimensions (e.g. Z) against externally provided rasters (e.g. DEMs) in the HAG filter. I know a lot of people who use this approach in their workflows with (non-lidar) point clouds.

**Comment by hobu** [comment]
> computing dimensions (e.g. Z) against externally provided rasters (e.g. DEMs) 

You have the ability to do this now with the filters.dem https://pdal.io/stages/filters.dem.html but it doesn't work inside the context of the HAG filter.

**Comment by AndrewCunliffe** [comment]
> 
> 
> > computing dimensions (e.g. Z) against externally provided rasters (e.g. DEMs)
> 
> You have the ability to do this now with the filters.dem https://pdal.io/stages/filters.dem.html but it doesn't work inside the context of the HAG filter.

Thanks @hobu, I had overlooked that option.

**Comment by AndrewCunliffe** [comment]
> 
> 
> > computing dimensions (e.g. Z) against externally provided rasters (e.g. DEMs)
> 
> You have the ability to do this now with the filters.dem https://pdal.io/stages/filters.dem.html but it doesn't work inside the context of the HAG filter.

Thanks @hobo. I'm sorry I probably I wasn't clear enough, but I was thinking of a filter to compute height-above-ground relative to a supplied DEM, rather than classified points or an-on-the fly TIN (but I'm looking forward to testing the new code @gadomski is developing). I think this is a bit different to filters.dem.

**Comment by abellgithub** [comment]
Closed with #2520
